lms/djangoapps/courseware/middleware.py

```diff
@@ -22,3 +25,21 @@ def process_exception(self, _request, exception):
                     args=[course_key.to_deprecated_string()]
                 )
             )
+
+
+class RedirectExpiredCoursesMiddleware(object):
+
+    def process_request(self, request):
+        if is_request_in_themed_site():
+            redirect_from = ''.join(
+                ('http', ('', 's')[request.is_secure()], '://', request.META['HTTP_HOST'], request.path)
+            )
+            cache_key = "redirect.{url}".format(url=redirect_from)
+            redirect_to = cache.get(cache_key, '')
+            if not redirect_to:
+                redirect_object = Redirect.objects.filter(old_path=redirect_from)
+                if len(redirect_object) > 0:
+                    redirect_to = redirect_object[0].new_path
+                    cache.set(cache_key, redirect_to, settings.REDIRECT_URL_CACHE_TIMEOUT)
+            if redirect_to:
+                return redirect(redirect_to)
```

Comment:
We should cache redirects by host. We can query the `Redirect` model using the `icontains` field lookup.
```
host = request.META['HTTP_HOST']
cache_key = "theme_redirect.{host}".format(host=host)
host_redirects = cache.get(cache_key)
if host_redirects is None:
    host_redirects = {redirect.old_path: redirect.new_path for redirect in Redirect.objects.filter(old_path__icontains=host)}
    cache.set(cache_key, host_redirects, settings.THEME_REDIRECT_CACHE_TIMEOUT)

redirect_to = host_redirects.get(redirect_from)
if redirect_to:
    return redirect(redirect_to)
```